src/Shared/ServerInfrastructure/BufferExtensions.cs

```diff
@@ -140,123 +133,32 @@ private static void WriteNumericMultiWrite(ref this BufferWriter<PipeWriter> buf
-            int i = 0;
-            // Use Intrinsic switch
-            if (IntPtr.Size == 8) // 64 bit
-            {
-                if (length < 4) goto trailing;
-
-                int unaligned = (int)(((ulong)input) & 0x7) >> 1;
-                // Unaligned chars
-                for (; i < unaligned; i++)
-                {
-                    char ch = *(input + i);
-                    *(output + i) = (byte)ch; // Cast convert
-                }
-
-                // Aligned
-                int ulongDoubleCount = (length - i) & ~0x7;
-                for (; i < ulongDoubleCount; i += 8)
-                {
-                    ulong inputUlong0 = *(ulong*)(input + i);
-                    ulong inputUlong1 = *(ulong*)(input + i + 4);
-                    // Pack 16 ASCII chars into 16 bytes
-                    *(uint*)(output + i) =
-                        ((uint)((inputUlong0 * Shift16Shift24) >> 24) & 0xffff) |
-                        ((uint)((inputUlong0 * Shift8Identity) >> 24) & 0xffff0000);
-                    *(uint*)(output + i + 4) =
-                        ((uint)((inputUlong1 * Shift16Shift24) >> 24) & 0xffff) |
-                        ((uint)((inputUlong1 * Shift8Identity) >> 24) & 0xffff0000);
-                }
-                if (length - 4 > i)
-                {
-                    ulong inputUlong = *(ulong*)(input + i);
-                    // Pack 8 ASCII chars into 8 bytes
-                    *(uint*)(output + i) =
-                        ((uint)((inputUlong * Shift16Shift24) >> 24) & 0xffff) |
-                        ((uint)((inputUlong * Shift8Identity) >> 24) & 0xffff0000);
-                    i += 4;
-                }
-
-                trailing:
-                for (; i < length; i++)
-                {
-                    char ch = *(input + i);
-                    *(output + i) = (byte)ch; // Cast convert
-                }
-            }
-            else // 32 bit
+            var dataLength = data.Length;
+            var offset = 0;
+            var bytes = buffer.Span;
+            do
             {
-                // Unaligned chars
-                if ((unchecked((int)input) & 0x2) != 0)
-                {
-                    char ch = *input;
-                    i = 1;
-                    *(output) = (byte)ch; // Cast convert
-                }
-
-                // Aligned
-                int uintCount = (length - i) & ~0x3;
-                for (; i < uintCount; i += 4)
+                var writable = Math.Min(dataLength - offset, bytes.Length);
+                // Zero length spans are possible, though unlikely.
+                // ASCII.GetBytes and .Advance will both handle them so we won't special case for them.
+                Encoding.ASCII.GetBytes(data.AsSpan(offset, writable), bytes);
+                buffer.Advance(writable);
+
+                // The `add` will macro-op fuse with `if` test.
```

Comment:
Out of curiosity, what architectures does this apply to? I assume Intel x86-64 CPUs do the macro-op fusion, but what about other CPUs?

Reply:
Intel has a 2000 patent on it [which expires in 2022-07-03](https://patents.google.com/patent/US6675376) so it will probably become more wide spread then 😅

Reply:
Comment's weird. How could and `add` instruction fuse with the `if` test which will be a `cmp`? Besides, why does it matter?

Reply:
Bleh, comment was from intermediate commit where I was justifying moving some stuff out of the `> 0` if statement.

Whereas just removed that `if` entirely.

Reply:
e.g. it didn't mix with the `if` test if it was inside a previous `if`

Reply:
Removed